GUI: More expressive widget icon colors, and more along the way

tangkong opened the pull request:
## Description
We're finally getting into the issues from the jira tickets!  Kinda.

- Gives the GUI a title that is not "form"
- Removes sorting of endstations to allow lazy loading of BeamPaths
- Fixes device filter area to match with `isinstance`, and no longer use factory functions from pcdsdevices. 
  - closes #144 
- Track cumulative transmission along BeamPath, and mark devices as blocking if they force total transmission below a threshold
- Color the widget based on both the device state and path state.   
  - A variety of jira tickets, including [jira1](https://jira.slac.stanford.edu/browse/LCLSECSD-143), [jira2](https://jira.slac.stanford.edu/browse/LCLSPC-135), [jira3](https://jira.slac.stanford.edu/browse/LCLSECSD-748)

Also in a more minor sense
- Fixes test suite fa4 deprecation warnings
- updates some docstrings and type hints
- updates tests
  - tests for attenuation tracking and widget color updates
  - un-hard-codes `path.show_devices()` test, using regex and pain
- makes the lcls simulated facility more accessible

## Motivation and Context
closes #144 , and a few jira tickets as detailed above

<details>
  <summary>Thoughts and justifications of my decisions, to make myself feel better</summary>
  
- Accumulated transmission logic: Inserted devices can now have non-0 transmission, but should removed devices have the ability to block?
  - Currently the logic already assumes that removed devices cannot be blocking, unless there is branch mismatch
  - While it's not absurd to imagine a device that blocks when removed, I chose to punt the issue
- Now that the `LightRow` widget relies on the path and the device, are we going to have a race condition?
  - Two `LightRow` widgets and one `BeamPath` are subscribed to changes in `device.lightpath_summary`.
  - When `lightpath_summary` changes, is it possible that the `LightRow`'s update before the `BeamPath` finishes, giving us bad state information?
  - No, callbacks are processed in insertion-order, and the `BeamPath` callbacks are subscribed before the `LightRow` widgets
  - I should really draw out the subscription structure in a nice way so people don't have to go through the maze repeatedly

</details>

## How Has This Been Tested?
Interactive testing, and tests have been added to the suite.

## Where Has This Been Documented?
This PR, some docstrings, etc

## Some example screenshots
From sim mode, sim_ipimb's have an inserted transmission of 0.5, so inserting 4 drops the beam below the minimum of 0.1.  Inserted devices are now blue, and device types have been updated
<img width="1546" alt="image" src="https://user-images.githubusercontent.com/35379409/185447515-0ad0bebc-0754-4a98-84f4-75133820706f.png">

An example from XCS
<img width="1357" alt="image" src="https://user-images.githubusercontent.com/35379409/185448801-faafca3d-8b7b-4ee7-8d0a-1be67d94dc6d.png">

tangkong commented:
This is probably ready for some eyes. The exact names and colors I used will surely be changed in the future, but hopefully this PR provides us an opportunity to customize such things.

I had thoughts of adding the option to switch between various color maps, but that's maybe for a different pr.

ZLLentz commented:
I'll give this a look! Any special instructions for running it myself? Do I need to set up a pcdsdevices dev too?

ZLLentz commented on lightpath/path.py:
```diff
@@ -214,12 +214,14 @@ def blocking_devices(self) -> List[Device]:
         Returns
         -------
         List[Device]
-            list of blockoing devices
```
This old typo gave me a giggle

ZLLentz commented on lightpath/path.py:
```diff
@@ -234,13 +236,17 @@ def blocking_devices(self) -> List[Device]:
             # check inserted
             elif curr_state is DeviceState.Inserted:
-                # Ignore devices with low enough transmssion
+                current_transmission *= curr_status.transmission
```
Do we need a check here for `curr_status.transmission` <= 1? Maybe an exception or a min(transmission, 1) clamp?

ZLLentz commented on lightpath/path.py:
```diff
@@ -234,13 +236,17 @@ def blocking_devices(self) -> List[Device]:
-            # Find unknown and inconsistent devices
+                # Any device seeing current transmission below min would block
+                elif current_transmission < self.minimum_transmission:
+                    block.append(device)
```
Does this new total transmission condition make the old single device condition redundant?

Note that this is a change in behavior from previous- I think it's more useful this way

tangkong replied:
I think this is also leftover from when I colored downstream half-removed devices as half-removed, not blocking.  It is redundant now.

ZLLentz commented on lightpath/ui/gui.py:
```diff
@@ -173,6 +172,40 @@ def hidden_devices(self):
+    def update_device_types(self):
+        """
+        Update the device type filter check boxes bases on devices
+        currently in the path
```
This is a good idea, it was always weird before having useless filters loaded

ZLLentz commented on lightpath/ui/gui.py:
```diff
@@ -44,10 +48,14 @@ class LightApp(Display):
+                   'PulsePicker': dtypes.PulsePicker, 'Slit': dtypes.Slits,
+                   'Stopper': dtypes.Stopper, 'XFLS': dtypes.XFLS,
+                   'SIM_IPIMB': IPIMB, 'SIM_Stopper': Stopper,
+                   'SIM_Crystal': Crystal}
```
Do we have to consider the placeholder items here for the devices that didn't load properly? Or is that baked in already?

tangkong replied:
Ah, I didn't include those.  I can do that.

tangkong replied:
Actually, I didn't forget it before (I just forgot to write down why I made that decision).  Because all the sim devices inherit from `Valve`, unchecking it removes all the sim devices.  While in sim-mode you can just put the devices back, in normal mode we mock devices with either `Valve` or `Crystal`, the latter of the two inheriting from `Valve`.  🤔

ZLLentz replied:
But surely there should be a way in non-sim mode to hide the mock devices? Either with their own button or with the original button for that class?

tangkong replied:
Agreed, I just need to change the mock classes so Crystal isn't also Valve

ZLLentz commented on lightpath/ui/gui.py:
```diff
@@ -274,6 +309,9 @@ def update_path(self, *args,  **kwargs):
                     widget.update_light(_in, _out)
+                    # Reconsider blocking device state
+                    if device is block:
+                        widget.update_state()
```
Is this an optimization of some sort? Or is it a bugfix?

tangkong replied:
Ah this was actually left over from when I colored downstream half-removed devices as half-removed, not blocking.  However this does remind me of a bug I'll fix shortly:

If you insert a bunch of attenuating devices, the last device gets colored red as expected.  But if you remove an upstream device and the final blocking device changes, the previous blocking device needs to change.

ZLLentz commented on lightpath/ui/widgets.py:
```diff
@@ -18,12 +18,15 @@
-                QColor(255, 0, 255)]  # Error
+state_colors = {
+    'removed': QColor(124, 252, 0),  # Removed (green)
+    'half_removed': QColor(0, 176, 255),  # half-removed (light blue)
```
Is this the "inserted but not blocking" state?

tangkong replied:
Yep, I'm sure there's a better name but that's what I went with for now.

ZLLentz replied:
I also have no better suggestion despite it seeming like one should exist

ZLLentz commented on lightpath/ui/widgets.py:
```diff
@@ -177,21 +180,52 @@ def add_signal(self, signal):
+        If device is an impediment: StateColor.Blocking
+        If device is removed: StateColor.Removed
+        If device is inserted, not blocking (mirrors): StateColor.HalfRemoved
+        If device is unknown / errored: StateColor.Error
```
Docstring here is left over from a previous iteration of the code where the state colors were an enum, I think?

tangkong replied:
Ah right, they're dictionary access now.

ZLLentz reviewed (commented):
This looks great to me
Added questions and comments to make sure every case is considered

tangkong commented:
> I'll give this a look! Any special instructions for running it myself? Do I need to set up a pcdsdevices dev too?

You'll need updates from 3 branches:
* The most recent `lightpath` branch/PR
* The LightpathMixin `pcdsdevices` PR: https://github.com/pcdshub/pcdsdevices/pull/1028
* The updated `device_config` implementing LightpathMixin on the happi database: https://github.com/pcdshub/device_config/pull/31

Once those three are in your path, you should be able to redirect `HAPPI_CFG` to the new device_config db and load lightpath.  Loading `lightpath --sim` should work independent of that

ZLLentz commented:
It looks to me like I also need to include `lightpath` as a pip -e install to make the container work right and I need to include the master branch of `happi` too

ZLLentz commented:
I get strange behavior when I run this: lots of error spam and then included device types is surprisingly bare. I suspect maybe I haven't set up my environment correctly but I can't tell. The new colors show up and look good.
![image](https://user-images.githubusercontent.com/10647860/185271085-6c34007c-fa5a-420b-8a29-c4b5a4f430e9.png)

tangkong commented:
Yeah, I'm working out the fix to the error spam now.  It's been bothering me for quite a while, and I think I made a breakthrough (stay tuned).

The device types list is bare because we're looking for specific pre-determined device types, and pcdsdevices has become quite... varied in their inheritance.  (e.g., at2l0 doesn't inherit from `AttBase`).  Something I wanted to do was populate that table with the existing devices, but couldn't immediately think of a clean way to decide what the "parent" of a device should be, or when to stop grouping devices together (a lot of devices might be reasonably grouped as "positioners", but not all of them?).  We could also let the user configure this, but we'd have to build up the machinery for that.  I remember thinking about this and eventually also punting it. 

If all this looks good I can make an issue

ZLLentz commented:
> The device types list is bare because we're looking for specific pre-determined device types, and pcdsdevices has become quite... varied in their inheritance.

Maybe this should become its own issue if there is enough unrelated stuff in this PR. I'd expect that I'd be able to make all of the devices disappear by clicking all the filter buttons, but that doesn't seem to be possible (it wasn't possible in the old lightpath either, to be fair)

`lightpath --sim` doesn't load properly for me which makes me think I've made an error in my dev setup here...

ZLLentz commented:
I don't think this closes #144 though. The buttons still don't work fully, there are still valves that remain when clicking the valves button for example.

ZLLentz commented on lightpath/ui/gui.py:
```diff
@@ -44,10 +48,14 @@ class LightApp(Display):
-                   dtypes.LODCM, dtypes.OffsetMirror, dtypes.PIM, PPSStopper,
-                   dtypes.PulsePicker, dtypes.Slits, dtypes.Stopper,
-                   dtypes.XFLS]
+    shown_types = {'Attenuator': AttBase, 'Gate Valve': dtypes.GateValve,
```
`isinstance` can check against tuples of types, so we could include tuples of types in the dict here to catch cases where multiple unrelated type trees represent the same "kind" of object.

Alternatively, maybe it's time to rework pcdsdevices to include bare `AttenuatorType` base classes with no code in them.

klauer replied:
I'd be OK with that or just type-hint based ones: `AttenuatorType = Union[AttBase, ...]`

klauer replied:
This could be dynamically generated based on the device instances that are currently shown, though it wouldn't be as nice as a manually curated list.

tangkong replied:
Zach and I settled on taking the module name as the "device type", which simplifies the work here and helps incorporate future changes more easily.  

Under this scheme, we'd group all devices from `pcdsdevices.attenuators` under the same checkbox, everything under `pcdsdevices.pim` together, etc

(I'll commit when I'm able, but do you have any objections to this approach?)

klauer replied:
Sounds good to me 👍

tangkong commented:
> I don't think this closes #144 though. The buttons still don't work fully, there are still valves that remain when clicking the valves button for example.

Which valves remain?  the SIM_Valve only refers to the mock device, and I think not all valves are GateValves?

ZLLentz commented:
Pretty much all valves in the lightpath are gate valves

ZLLentz commented:
From this screenshot:

tv1l0_vgc01 should be grouped with the gate valves (class `VGC` = Valve of Gate type Controlled)
im1l0 should be grouped with the pims (class `XPIM` = XTES PIM)
sl1l0 should be grouped with the slits (class `PowerSlits`)

![image](https://user-images.githubusercontent.com/10647860/185468661-ac13dedb-5574-41bf-ae92-c37887a36a5d.png)

I think we should sit down and try to put all the remaining beamline objects into one of the existing boxes

ZLLentz commented:
Lingering combobox error spam can be sourced to https://github.com/pcdshub/pcdsdevices/pull/1046

klauer commented on lightpath/ui/gui.py:
```diff
@@ -96,39 +107,28 @@ def __init__(self, controller, beamline=None,
         if dark:
             typhos.use_stylesheet(dark=True)
 
+        self.setWindowTitle('Lightpath')
```
Title change, yes! 😁 

```suggestion
        self.setWindowTitle(f'Lightpath - {beamline}' if beamline else "Lightpath")
```

klauer commented on lightpath/ui/gui.py:
```diff
@@ -173,6 +173,40 @@ def hidden_devices(self):
+            for child in self._device_checkboxes:
+                self.device_types.layout().removeWidget(child)
+                self.device_buttons.pop(child)
+                child.deleteLater()
```
I know there was talk of a double free-related segfault. I admittedly didn't pay enough attention to know if it could be related, but I see `deleteLater` here...

You could avoid deleting these by just creating all of the checkboxes and hiding the ones that aren't useful to avoid a create/remove/repeat

tangkong replied:
I could make it so we never actually delete these checkboxes, rather we just store them and pick the ones we need.  

Is it generally bad to `deleteLater`?  I figured it'd help free up memory and guard against the strange off chance that we accumulate too many device types, but I could try to avoid it in general.

klauer replied:
If I hadn't vaguely recalled about the double free, I wouldn't have even commented. Did you by chance get to the bottom of that issue?

Normally `deleteLater` is just fine and isn't a reason on its own to suggest not using it.  There are scenarios where the overhead of creating/deleting/creating can be worse than just creating everything and hiding what you don't need, though it on a case-by-case basis where either method would be faster/more efficient/preferable.

tangkong replied:
I poked around a bunch trying to reproduce the error, but to no avail.  

Just wanted to make sure I wasn't missing anything canonical w.r.t. pyqt gui development, and I'll certainly keep an eye out for the double free errors

klauer commented on lightpath/ui/widgets.py:
```diff
@@ -177,21 +180,52 @@ def add_signal(self, signal):
+
+        if device_state is DeviceState.Disconnected:
+            return state_colors['disconnected']
+        elif device_state is DeviceState.Error:
```
Minor style suggestion - big blocks containing multiple `elif` aren't really necessary when there's `return` in each of the clauses:
```suggestion
        if device_state is DeviceState.Error:
```
Then `else` even becomes unnecessary

tangkong commented:
I think I've addressed the concerns here.  I spent this morning poking around the gui trying to illicit a "double free" error, but to no avail.  Instead I:

- found one small bug that was fixed by removing code that modified checkboxes outside of the new `update_device_types` method
- ran into a segmentation fault I could also not replicate

ZLLentz commented:
Do you want a re-review and/or a re-beta tester?

tangkong commented:
> Do you want a re-review and/or a re-beta tester?

A re-review would be nice, the main thing that's changed is the device types filtering.  Someone else poking around with it would also be appreciated, maybe the sporadic errors/crashes will show up for others...

Once I merge this I have a documentation rewrite in the pipeline, after which maybe we can tag? 👀

ZLLentz commented:
Looks good interactively! I'll poke through the source now.
There are a few devices that filter on an unintuitive button but I think that means they need happi/pcdsdevices updates rather than anything special here.

ZLLentz commented:
I hit a segmentation fault but it wasn't the double free. I'll run again and turn core dumping on while I browse the code.

ZLLentz commented on lightpath/ui/widgets.py:
```diff
@@ -177,21 +180,52 @@ def add_signal(self, signal):
+        QColor
+            the color to apply to the icon
+        """
+        device_state, _ = find_device_state(self.device)
```
I'm not 100% sure about this, but I think this bit might make the lightpath first load go painfully slow when devices are disconnected?

Something is pathing through https://github.com/pcdshub/lightpath/blob/05a9394e09acf256649738f6233d54a7ba2b1b4d/lightpath/path.py#L83 at startup for all devices

Might be good to look at this in a follow-up

ZLLentz replied:
What I'm trying to say is: all disconnected devices are timing out at one of these function calls in series before any device is shown in lightpath

ZLLentz replied:
It might worse than this actually, I can't seem to switch to a path that has a disconnected device
I just see this forever and my ui freezes
```
[2022-08-24 16:36:19] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:21] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:29] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:31] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:39] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:41] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:49] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:51] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:36:59] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:01] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:09] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:11] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:19] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:21] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:29] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:31] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:39] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:41] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:49] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:51] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:37:59] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:38:01] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:38:09] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:38:11] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
[2022-08-24 16:38:19] - WARNING -  Unable to connect to BeckhoffSlits(SL2K2:SCATTER, name=sl2k2)
```

tangkong replied:
This makes sense.  I think the real root of the issue is the fact that I increased the default EpicsSignal timeout to 10s, otherwise we'd get a slew of timeout errors at startup.  

Each `LightRow` widget updates its state once after they are created, which calls `find_device_state`.  

Something I considered doing was simply making timeout errors less severe, and assuming that devices would connect eventually.  That seemed dangerous though.

ZLLentz replied:
> Each LightRow widget updates its state once after they are created, which calls find_device_state.

Somehow there must be at least one other place this is done that blocks qt from doing anything forever

tangkong commented:
If there are no further comments I'll merge this at end of day (8/25/22).  The disconnected devices issue will be a bit involved, and deserves a separate PR.

ZLLentz commented:
I agree, this is a strong improvement and fixes a lot!
